java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,354 @@
+    }
+
+    try {
+      SecretKeySpec keySpec = new SecretKeySpec(key, "HmacSHA256");
+      mac.init(keySpec);
+    } catch (InvalidKeyException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    return mac;
+  }
+
+  /**
+   * Generates a message for MAC by concatenating shared secret and two keys.
+   *
+   * @param sharedSecret some shard secret.
+   * @param receiverPublicKey receiver's public key.
+   * @param senderPublicKey sender's public key.
+   * @return generated MAC message.
+   */
+  private byte[] generateMacMsg(byte[] sharedSecret, byte[] receiverPublicKey,
+      byte[] senderPublicKey) {
+    return ByteBuffer
+        .allocate(sharedSecret.length + receiverPublicKey.length + senderPublicKey.length)
+        .put(sharedSecret)
+        .put(receiverPublicKey)
+        .put(senderPublicKey)
+        .array();
+  }
+
+  /**
+   * Generates a tag corresponding to the given message {@code msg} and the counter {@code cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @param mac MAC key of a sender.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   */
+  private byte[] generateTag(byte[] msg, int cnt, Mac mac) throws CounterOutOfBoundsException {
+    if (cnt < MIN_CNT_VALUE || cnt > MAX_CNT_VALUE) {
+      throw new CounterOutOfBoundsException(
+          String.format("Counter should be in [0..255] range. Got %d.", cnt));
+    }
+
+    byte[] finalMsg = new byte[msg.length + 1];
+    finalMsg[0] = (byte) cnt; // for [0..255] range do `& 0xFF`
+    arraycopy(msg, 0, finalMsg, 1, msg.length);
+
+    return mac.doFinal(finalMsg);
+  }
+
+  /**
+   * Exception, which is thrown whenever a counter is out of [0..255] range.
+   */
+  public static class CounterOutOfBoundsException extends Exception {
+
+    CounterOutOfBoundsException(String msg) {
+      super(msg);
+    }
+
+  }
+
+  /**
+   * Exception, which is thrown whenever a minimum tag length is out of [1..32] range.
+   */
+  public static class MinPeerTagLengthOutOfBoundsException extends Exception {
+
+    MinPeerTagLengthOutOfBoundsException(String msg) {
+      super(msg);
+    }
+
+  }
+
+  /**
+   * Exception, which is thrown whenever the received tag has invalid length or it's not equal to
+   * the prefix of a correct tag.
+   */
+  public static class WrongTagException extends Exception {
+
+    WrongTagException(String msg) {
+      super(msg);
+    }
+
+  }
+
+  /**
+   * Provides static private key, so doesn't generate a new random one.
+   */
+  public static class StaticSecureRandom extends SecureRandom {
+
+    private final byte[] privateKey;
+
+    /**
+     * Stores the given private key.
+     *
+     * @param privateKey private key to be stored.
+     */
+    public StaticSecureRandom(Optional<byte[]> privateKey) {
+      this.privateKey = privateKey.map(byte[]::clone).orElse(null);
```

Comment:
This class accepts a null argument, but would then NPE when nextBytes is called.

Reply:
Not actual anymore